File: src/kinematics_rb1/odometry_reader.cpp
#include "kinematics_rb1/odometry_reader.hpp"
#include "tf/LinearMath/Matrix3x3.h"

void OdometryReader::odom_subscriber_cb(const nav_msgs::Odometry::ConstPtr &msg) {
    odom_pose["x"] = msg->pose.pose.position.x;
    odom_pose["y"] = msg->pose.pose.position.y;
    tf::Quaternion q = tf::Quaternion(msg->pose.pose.orientation.x, msg->pose.pose.orientation.y, msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);
    tf::Matrix3x3 m(q);
    double roll, pitch, yaw;
    m.getRPY(roll, pitch, yaw);
    odom_pose["theta"] = yaw;
}

OdometryReader::OdometryReader(ros::NodeHandle &nh, const std::string &topic) : nh_(nh) {
  odom_subscriber = nh_.subscribe<nav_msgs::Odometry>(topic, 100, &OdometryReader::odom_subscriber_cb, this);
  ros::Rate rate(10);
  rate.sleep();
}

void OdometryReader::unregister(void) {
  odom_subscriber.shutdown();
}

File: scripts/odometry_reader.py
#! /usr/bin/env python

import rospy
import numpy as np
from nav_msgs.msg import Odometry
from tf.transformations import euler_from_quaternion

class OdometryReader():
    def __init__(self, topic):
        self.odom_pose = {}
        self.trajectory = []
        self.topic = topic
        self.subscribe()

    def callback(self, msg):
        self.odom_pose['x'] = msg.pose.pose.position.x
        self.odom_pose['y'] = msg.pose.pose.position.y
        self.trajectory.append((self.odom_pose['x'], self.odom_pose['y']))
        (_, _, self.odom_pose['theta']) = euler_from_quaternion([msg.pose.pose.orientation.x, 
                                                            msg.pose.pose.orientation.y, 
                                                            msg.pose.pose.orientation.z, 
                                                            msg.pose.pose.orientation.w])
        rospy.loginfo(self.odom_pose)

    def subscribe(self):
        self.odom_subscriber = rospy.Subscriber(self.topic, Odometry, self.callback)
        rospy.sleep(0.1)

    def unregister(self):
        #np.save('trajectory',self.trajectory)
        self.odom_subscriber.unregister()

if __name__ == "__main__":
    rospy.init_node('odom_reader')
    rospy.loginfo("Odometry node initialized.")
    oc = OdometryReader("/rb1_odom")
    rospy.spin()